core/src/main/scala/kafka/log/Log.scala

```diff
@@ -321,33 +326,37 @@ class Log(@volatile var dir: File,
     *
     * @param lastOffset
     *
-    * @return An instance of IdMapping
+    * @return An instance of ProducerIdMapping
     */
-  private def buildAndRecoverIdMap(lastOffset: Long): IdMapping = {
-//    val newIdMap = new IdMapping(topicAndPartition, config, dir, logStartOffset, lastOffset)
-//    // Iterate over the messages between idMap.lastSnapshotOffset
-//    // and the last offset of the active segment
-//    val msgs = read(idMap.mapEndOffset, 0, Some(logEndOffset))
-//    msgs.logBuffer.asScala.foreach(mao => {
-//      mao.firstOffset
-//    })
-//    newIdMap
-    null
+  private def buildAndRecoverPidMap(lastOffset: Long) {
+    lock synchronized {
+      pidMap.truncateTo(lastOffset)
```

Comment:
Would be useful to add a comment saying that `truncateTo` actually loads the last snapshot into memory as well. The name doesn't suggest this.